styx-client/src/main/java/com/spotify/styx/client/StyxApolloClient.java

```diff
@@ -0,0 +1,319 @@
+
+/**
+ * Styx Apollo Client Implementation
+ */
+public class StyxApolloClient
+    implements StyxWorkflowClient, StyxBackfillClient, StyxResourceClient,
+               StyxSchedulerClient, StyxStatusClient {
+  private static final String STYX_API_ENDPOINT = "/api/v2";
+  private static final String UTF_8 = "UTF-8";
+  private static final int TTL_SECONDS = 90;
+
+  private final Client client;
+  private final String apiHost;
+  private final String userAgentIdentifier;
+
+  public StyxApolloClient(final Client client,
+                          final String apiHost,
+                          final String userAgentIdentifier) {
+    this.client = client;
+    this.apiHost = apiHost;
+    this.userAgentIdentifier = userAgentIdentifier;
+  }
+
+  @Override
+  public CompletionStage<RunStateDataPayload> activeStates(Optional<String> componentId) {
+    String url = apiUrl("status", "activeStates");
+    if (componentId.isPresent()) {
+      url = addQueryToApiUrl(url, "component=" + componentId.get());
+    }
+    return executeRequest(
+        Request.forUri(url).withTtl(Duration.ofSeconds(TTL_SECONDS)),
+        RunStateDataPayload.class);
+  }
+
+  @Override
+  public CompletionStage<List<EventInfo>> eventsForWorkflowInstance(String componentId,
+                                                                    String workflowId,
+                                                                    String parameter) {
+    final String url = apiUrl("status", "events", componentId, workflowId, parameter);
+    return executeRequest(Request.forUri(url).withTtl(Duration.ofSeconds(TTL_SECONDS)))
+        .thenApply(response -> {
+          final JsonNode jsonNode;
+          try {
+            if (!response.payload().isPresent()) {
+              throw new RuntimeException("No json returned from API");
+            }
+            jsonNode = OBJECT_MAPPER.readTree(response.payload().get().toByteArray());
+          } catch (IOException e) {
+            throw new RuntimeException("Invalid json returned from API");
+          }
+
+          if (!jsonNode.isObject()) {
+            throw new RuntimeException("Unexpected json returned from API");
+          }
+
+          final ObjectNode json = (ObjectNode) jsonNode;
+          final ArrayNode events = json.withArray("events");
+          final ImmutableList.Builder<EventInfo> eventInfos = ImmutableList.builder();
+          for (JsonNode eventWithTimestamp : events) {
+            final long ts = eventWithTimestamp.get("timestamp").asLong();
+            final JsonNode event = eventWithTimestamp.get("event");
+
+            String eventName;
+            String eventInfo;
+            try {
+              Event typedEvent = OBJECT_MAPPER.convertValue(event, Event.class);
+              eventName = EventUtil.name(typedEvent);
+              eventInfo = EventUtil.info(typedEvent);
+            } catch (IllegalArgumentException e) {
+              // fall back to just inspecting the json
```

Comment:
I think it is fine to be relaxed in the client for this API as well, since `EventInfo` is mainly "informational" thus it feels ok to make a best effort in constructing it. This detail about forward compatibility can be mentioned in the Javadocs though, I will add it.